pkg/agent/runner.go

```diff
@@ -322,6 +333,102 @@ func (r *Runner) Run(ctx context.Context, logger *zap.Logger, vmInfoUpdated util
 	}
 }
 
+func (r *Runner) reportScalingEvent(timestamp time.Time, currentCU, targetCU uint32) {
+	var endpointID string
+
+	enabled := func() bool {
+		r.status.mu.Lock()
+		defer r.status.mu.Unlock()
+
+		endpointID = r.status.endpointID
+		return endpointID != "" && r.status.vmInfo.Config.ReportScalingEvents
+	}()
+	if !enabled {
+		return
+	}
+
+	reporter := r.global.scalingReporter
+	reporter.Submit(reporter.NewRealEvent(
+		timestamp,
+		endpointID,
+		currentCU,
+		targetCU,
+	))
+}
+
+func (r *Runner) reportDesiredScaling(
+	rl *desiredScalingReportLimiter,
+	timestamp time.Time,
+	currentCU uint32,
+	targetCU uint32,
+	parts scalingevents.GoalCUComponents,
+) {
+	var endpointID string
+
+	enabled := func() bool {
+		r.status.mu.Lock()
+		defer r.status.mu.Unlock()
+
+		endpointID = r.status.endpointID
+		return endpointID != "" && r.status.vmInfo.Config.ReportDesiredScaling
+	}()
+	if !enabled {
+		return
+	}
+
+	// TODO: Use this opportunity to report the desired scaling in the per-VM
+	// metrics.
+
+	rl.report(r.global.scalingReporter, timestamp, endpointID, currentCU, targetCU, parts)
+}
+
+type desiredScalingReportLimiter struct {
+	lastTarget *uint32
+	lastParts  *scalingevents.GoalCUComponents
+}
+
+func (rl *desiredScalingReportLimiter) report(
```

Comment:
Potential alternative would be to have 
```go
type Skipper interface {
	Skip(event ScalingEvent) bool
}
```
injected into `Reporter`, and it would be called in `Submit()`.

This would allow to implement limiter in a more generic way, for any type of event. Plus no need to pass so many arguments, when we can pass only `ScalingEvent`.

Reply:
Hm, that's a fair point. I agree the current situation is overly complex; I'm not sure that adding a `Skipper` interface makes it simpler.

I'll think about it and get back you here...